consensus/common/src/main/java/tech/pegasys/pantheon/consensus/common/jsonrpc/AbstractGetSignerMetricsMethod.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.consensus.common.jsonrpc;
+
+import tech.pegasys.pantheon.consensus.common.VoteTallyCache;
+import tech.pegasys.pantheon.ethereum.core.Address;
+import tech.pegasys.pantheon.ethereum.core.BlockHeader;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.JsonRpcRequest;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.parameters.JsonRpcParameter;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.queries.BlockchainQueries;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.response.JsonRpcError;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.response.JsonRpcErrorResponse;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.response.JsonRpcResponse;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.response.JsonRpcSuccessResponse;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.results.report.block.production.ProposerReportBlockProductionResult;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.results.report.block.production.SignerMetricsResult;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.results.report.block.production.ValidatorReportBlockProductionResult;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.LongStream;
+
+public abstract class AbstractGetSignerMetricsMethod {
+
+  private final BlockchainQueries blockchainQueries;
+  private final VoteTallyCache voteTallyCache;
+  private final JsonRpcParameter parameters;
+
+  public AbstractGetSignerMetricsMethod(
+      final BlockchainQueries blockchainQueries,
+      final VoteTallyCache voteTallyCache,
+      final JsonRpcParameter parameter) {
+    this.blockchainQueries = blockchainQueries;
+    this.voteTallyCache = voteTallyCache;
+    this.parameters = parameter;
+  }
+
+  public JsonRpcResponse response(final JsonRpcRequest request) {
+
+    final long startBlock = parameters.required(request.getParams(), 0, Long.class);
+    final long endBlock = parameters.required(request.getParams(), 1, Long.class);
+
+    if (isValidParameters(startBlock, endBlock)) {
+
+      final Map<Address, ProposerReportBlockProductionResult> proposersMap = new HashMap<>();
+      final Map<Address, ValidatorReportBlockProductionResult> validatorInLastBlockMap =
+          new HashMap<>();
```

Comment:
Hopefully the clarifications in my overall comment help to make sense of this but we'd only need the one map here.  The basic approach would be:

1. Iterate through each block in the range and increment the count of number of proposed blocks for the proposer of that block.  Note you might still get some proposers who don't yet exist in the map if they've been voted out before the end of the range.
2. For each validator in the last block of the range not already in `proposerMap`, add it with 0 proposed blocks. That will ensure they're all present in the report even if none ever propose blocks.